metals/src/main/scala/scala/meta/internal/parsing/Trees.scala

```diff
@@ -0,0 +1,102 @@
+package scala.meta.internal.parsing
+
+import scala.collection.concurrent.TrieMap
+
+import scala.meta._
+import scala.meta.internal.metals.Buffers
+import scala.meta.internal.metals.BuildTargets
+import scala.meta.internal.mtags.MtagsEnrichments._
+import scala.meta.parsers.Parsed
+
+import org.eclipse.lsp4j.Diagnostic
+import org.eclipse.lsp4j.DiagnosticSeverity
+
+/**
+ * Manages parsing of Scala source files into Scalameta syntax trees.
+ *
+ * - provides the latest good Scalameta tree for a given source file
+ *   similar as `Buffers` provides the current text content.
+ * - produces diagnostics for syntax errors.
+ */
+final class Trees(
+    buildTargets: BuildTargets,
+    buffers: Buffers,
+    defaultDialect: Dialect = dialects.Scala213
+) {
+
+  private val trees = TrieMap.empty[AbsolutePath, Tree]
+
+  def get(path: AbsolutePath): Option[Tree] =
+    trees.get(path).orElse {
+      // Fallback to parse without caching result.
+      parse(path).flatMap { _._1.toOption }
+    }
+
+  def tokenize(
+      path: AbsolutePath
+  ): Option[Tokens] = {
+    buffers.get(path).flatMap { text =>
+      val dial = dialect(path)
+      val input = Input.VirtualFile(path.toString(), text)
+      dial(input).tokenize.toOption
+    }
+  }
+
+  def didClose(fileUri: AbsolutePath): Unit = {
+    trees.remove(fileUri)
+  }
+
+  def didChange(path: AbsolutePath): List[Diagnostic] = {
+    parse(path) match {
+      // only publish parser diagnostics for Scala 2, which does not support significant indentation
+      case Some((parsed, dialect)) if !dialect.allowSignificantIndentation =>
+        parsed match {
+          case Parsed.Error(pos, message, _) =>
+            List(
+              new Diagnostic(
+                pos.toLSP,
+                message,
+                DiagnosticSeverity.Error,
+                "scalameta"
+              )
+            )
+          case Parsed.Success(tree) =>
+            trees(path) = tree
+            List()
+        }
+      case None =>
+        () // Unknown extension, do nothing.
+        List()
+    }
+  }
```

Comment:
So unless I'm missing where, it seems that the only place this is used is in `MetalsLanguageServer` in `buildServerPromise`, and we ignore the value. So I'm curious about this signature it it returning a `List[Diagnostic` when it seems that those are never actually used anywhere?

Reply:
Also as a total aside, this may benefit from a scaladoc. I feel like a lot of this may not be entirely clear from the outside what this method actually does when you just see `path -> List[Diagnostic]` as the signature.

Reply:
Ok, so actually this was a bug. We should have been doing something with the return value, I fixed it and added a scaladoc.